### PR TITLE
Support parameterized types in type/2

### PR DESCRIPTION
**Pull request by v0idpwn**
Related to discussion in #3784

### CODE REVIEW

File: lib/ecto/query/builder.ex [hunk 1]
```diff
@@ -83,6 +83,13 @@ defmodule Ecto.Query.Builder do
   end
 
   # tagged types
+  def escape({:type, x, [expr, parameterized_type, type_params]}, type, params_acc, vars, env) do
+    parameterized_type = Macro.expand(parameterized_type, get_env(env))
+    expanded_type = {:parameterized, parameterized_type, parameterized_type.init(type_params)}
+
+    escape({:type, x, [expr, expanded_type]}, type, params_acc, vars, env)
```
Comment (v0idpwn) [inline lib/ecto/query/builder.ex:90]:
```suggestion
  def escape({:type, meta, [expr, parameterized_type, type_params]}, type, params_acc, vars, env) do
    parameterized_type = Macro.expand(parameterized_type, get_env(env))
    expanded_type = {:parameterized, parameterized_type, parameterized_type.init(type_params)}

    escape({:type, meta, [expr, expanded_type]}, type, params_acc, vars, env)
```

**Comment by v0idpwn** [comment]
I don't understand why test is failing on CI, it runs fine on my machine... It has a compile-time dependency to test helpers, but I guess test helpers are available at the time the tests are compiled?

File: lib/ecto/query/api.ex [hunk 2]
```diff
@@ -645,9 +645,18 @@ defmodule Ecto.Query.API do
...
 
+  For parameterized types, use `type/3`
   """
   def type(interpolated_value, type), do: doc! [interpolated_value, type]
 
+  @doc """
+  Same as type/2, but for parameterized types.
+
+  from p in Post,
+    type(fragment("array_agg(?)", p.tag), {:array, Ecto.Enum}, values: [:a, :b])
```
Comment (v0idpwn) [inline lib/ecto/query/api.ex:656]:
```suggestion
      from p in Post,
        type(fragment("array_agg(?)", p.tag), {:array, Ecto.Enum}, values: [:a, :b])
```

File: lib/ecto/query/builder.ex [hunk 3]
```diff
@@ -682,6 +689,8 @@ defmodule Ecto.Query.Builder do
     do: type
   def validate_type!({:__aliases__, _, _} = type, _vars, env),
     do: Macro.expand(type, get_env(env))
+  def validate_type!({:parameterized, _, _} = type, _vars, _env),
+    do: Macro.escape(type)
```
Comment (v0idpwn) [inline lib/ecto/query/builder.ex:693]: In a second thought, I should've escaped this **before** the validate_type!/3 call, right?

**Comment by v0idpwn** [comment]
I get the reasoning and agree on removing the `type/3` expansion.

But to support the expanded parameterized type on `type/2` (as you suggested) we should merge part of this PR, as otherwise, it will raise a bit further in the pipeline.

**Comment by josevalim** [comment]
@v0idpwn sounds good to me! In the test, let's build the type using Ecto.PT.init/2 too instead of hardcoding it.

**Comment by v0idpwn** [comment]
I added a mention in type/2 docs, but didn't elaborate (and I'm not even sure if we should mention it), as it seems like this is not the first recommended approach

File: lib/ecto/query/api.ex [hunk 4]
```diff
@@ -618,6 +618,10 @@ defmodule Ecto.Query.API do
 
       type(^title, p.title)
 
+  Or a parameterized type, which must be already initialized:
+
+      type(^title, ^@parameterized_type)
```
Comment (josevalim) [inline lib/ecto/query/api.ex:623]:
```suggestion
  Or a parameterized type, which must be previously initialized
  with `Ecto.ParameterizedType.init/2`:

      @my_enum Ecto.ParameterizedType.init(Ecto.Enum, values: [:foo, :bar, :baz])
      type(^title, ^@my_enum)
```

**Review by josevalim — APPROVED** [review]
Just one tiny comment and ship it. :)

**Pull request merged**